pytorch/pytorch
# WIP: Fix build issues due to hardcoded /usr/local/cuda/lib64/libculibos.a

In the current libtorch nightlies, the file `libtorch/share/cmake/Caffe2/Caffe2Targets.cmake` contains:
```
INTERFACE_LINK_LIBRARIES "caffe2::cudart;c10_cuda;caffe2;caffe2::cufft;caffe2::curand;caffe2::cudnn;/usr/local/cuda/lib64/libculibos.a;dl;/usr/local/cuda/lib64/libculibos.a;caffe2::cublas"
```

So `libculibos.a` is included twice, with the full path.  That hardcoded `/usr/local/cuda` should instead be `${CUDA_TOOLKIT_ROOT_DIR}`. There are distributions (e.g. Arch Linux and derivatives) where the install location is `/opt/cuda` instead, and that results in users not
being able to build C++ extensions.

The root cause of the issue is that all dependencies were added to a list,  `Caffe2_PUBLIC_CUDA_DEPENDENCY_LIBS`. Adding strings to a list directly causes variable expansion.  The only way to preserve `${CUDA_TOOLKIT_ROOT_DIR}` seems to be to not use that list + `target_link_libraries`. Instead use the more standard `set_property(TARGET torch PROPERTY INTERFACE_LINK_LIBRARIES ...)`. A bit more verbose, but more robust.

To reproduce the issue:

1. download the latest `libtorch` nightly
2. Use, e.g., `example-app.cpp` from https://pytorch.org/tutorials/advanced/cpp_export.html/torch_script_custom_ops.html (follow steps 1,2,3). Your test directory looks like:
  `CMakeLists.txt  example-app.cpp  libtorch/  model.pt`
3. Use the following `build_example.sh` script:
```
mkdir build
cd build
cmake -DCMAKE_PREFIX_PATH=/home/rgommers/code/tmp/issue15476/libtorch ..
make
cd ..
```
4. `$ source build_example.sh` this will end with `No rule to make target '/usr/local/cuda/lib64/libculibos.a', needed by 'example-app'`.

Now with a `libtorch` binary produced with this fix included, that should work.

Note, I verified that removing `libculibos.a` completely from `Caffe2Targets.cmake` will fix the example app, but as explained by @pjh5 in https://github.com/pytorch/pytorch/issues/15476#issuecomment-515558702 that's only due to what `pytorch/builder` does. We can't leave out the dependency from the build scripts. This is the fix with minimal changes in the logic and no change in ordering of dependencies that I could come up with.

Closes gh-15476. Note, after merging I will go back and verify that the other issues that were closed as duplicates of gh-15476 are also fixed. Also note, I didn't manage to reproduce what `pytorch/builder` does locally exactly; if it's possible to get the build artifacts from the same CI job that produces the nightly before merging, that would be helpful. If not, this is small enough to review and be confident about it not breaking anything if CI passes.
